### ISSUE TITLE
TIMESTEP parameter of LMD files

### ISSUE THREAD

**Issue by jgarces02** [SamGG/flowAI#1]
Hi @giannimonaco, 

Thanks for so useful package. I'm using flow cytometry files from a Navios cytometer and final files are in LMD format (instead classic FCS ones). I'm having some problems with `TIMESTAMP` step because its `names(set[[1]]@description)` are different from standard FCS... this would be my values for this command, maybe the nearer could be `@ACQTIME`: 
```
[325] "$DFC7TO9"                     "$DFC8TO9"                     "$DFC9TO9"                     "$DFC10TO9"                   
[329] "$DFC1TO10"                    "$DFC2TO10"                    "$DFC3TO10"                    "$DFC4TO10"                   
[333] "$DFC5TO10"                    "$DFC6TO10"                    "$DFC7TO10"                    "$DFC8TO10"                   
[337] "$DFC9TO10"                    "$DFC10TO10"                   "@RATIONUMERATORMUX"           "@RATIODENOMINATORMUX"        
[341] "@REDLASERPOWER_START"         "@REDLASERPOWER_END"           "@REDLASERSHUTTER"             "@REDTARGETPOWER"             
[345] "@BLUELASERPOWER_START"        "@BLUELASERPOWER_END"          "@BLUELASERSHUTTER"            "@BLUETARGETPOWER"            
[349] "@VIOLETLASERPOWER_START"      "@VIOLETLASERPOWER_END"        "@VIOLETLASERSHUTTER"          "@VIOLETTARGETPOWER"          
[353] "@CYTOLINKMULTIPLECLIENT"      "@DISCRIMINATOR"               "@RATIO_NUMERATOR"             "@RATIO_DENOMINATOR"          
[357] "@RESAVEDFILE"                 "@BARCODE"                     "@CAROUSEL"                    "@TUBENO"                     
[361] "@PANEL"                       "@SETTINGSFILE"                "@SETTINGSFILEDATETIME"        "@SAMPLEID1"                  
[365] "@SAMPLEID2"                   "@SAMPLEID3"                   "@SAMPLEID4"                   "@ABSCALFACTOR"               
[369] "@TESTNAME"                    "@TESTFILE"                    "@CYTOMETERID"                 "@ACQTIME"                    
[373] "@ELAPSEDTIME"                 "@BUILDNUMBER"                 "$FIL"                         "@Acquisition Protocol Offset"
[377] "FILENAME"                     "GUID"                         "ORIGINALGUID"                 "GUID.original"   
```
Is there any way to indicate this "element" is changing? Thanks in advance!

**Comment by giannimonaco** [comment]
Hi,

I am sorry you got this problem. flowAI is supposed to work with files that are FCS 3.0 or 3.1, which is the standard format supported and mantained by the International Society for Advancement of Cytometry (ISAC).
At the moment, LMD files are not supported by flowAI. 

Best wishes,

Gianni



On Mon, 9 Mar 2020 at 16:58, jgarces <notifications@github.com <mailto:notifications@github.com>> wrote:
Hi @giannimonaco <https://github.com/giannimonaco>,

Thanks for so useful package. I'm using flow cytometry files from a Navios cytometer and final files are in LMD format (instead classic FCS ones). I'm having some problems with TIMESTAMP step because its names(set[[1]]@description) are different from standard FCS... this would be my values for this command, maybe the nearer could be @ACQTIME:

[325] "$DFC7TO9"                     "$DFC8TO9"                     "$DFC9TO9"                     "$DFC10TO9"                   
[329] "$DFC1TO10"                    "$DFC2TO10"                    "$DFC3TO10"                    "$DFC4TO10"                   
[333] "$DFC5TO10"                    "$DFC6TO10"                    "$DFC7TO10"                    "$DFC8TO10"                   
[337] "$DFC9TO10"                    "$DFC10TO10"                   "@RATIONUMERATORMUX"           "@RATIODENOMINATORMUX"        
[341] "@REDLASERPOWER_START"         "@REDLASERPOWER_END"           "@REDLASERSHUTTER"             "@REDTARGETPOWER"             
[345] "@BLUELASERPOWER_START"        "@BLUELASERPOWER_END"          "@BLUELASERSHUTTER"            "@BLUETARGETPOWER"            
[349] "@VIOLETLASERPOWER_START"      "@VIOLETLASERPOWER_END"        "@VIOLETLASERSHUTTER"          "@VIOLETTARGETPOWER"          
[353] "@CYTOLINKMULTIPLECLIENT"      "@DISCRIMINATOR"               "@RATIO_NUMERATOR"             "@RATIO_DENOMINATOR"          
[357] "@RESAVEDFILE"                 "@BARCODE"                     "@CAROUSEL"                    "@TUBENO"                     
[361] "@PANEL"                       "@SETTINGSFILE"                "@SETTINGSFILEDATETIME"        "@SAMPLEID1"                  
[365] "@SAMPLEID2"                   "@SAMPLEID3"                   "@SAMPLEID4"                   "@ABSCALFACTOR"               
[369] "@TESTNAME"                    "@TESTFILE"                    "@CYTOMETERID"                 "@ACQTIME"                    
[373] "@ELAPSEDTIME"                 "@BUILDNUMBER"                 "$FIL"                         "@Acquisition Protocol Offset"
[377] "FILENAME"                     "GUID"                         "ORIGINALGUID"                 "GUID.original"   
Is there any way to indicate this "element" is changing? Thanks in advance!

**Comment by giannimonaco** [comment]
Hi,

I am sorry you got this problem. flowAI is supposed to work with files that
are FCS 3.0 or 3.1, which is the standard format supported and mantained by
the International Society for Advancement of Cytometry (ISAC).
At the moment, LMD files are not supported by flowAI.

Best wishes,

Gianni

**Comment by jgarces02** [comment]
Yes, I agree that LMD is not the standard format... and any idea what parameter could be the equivalent to `TIMESTEP`? Thanks again.

**Comment by giannimonaco** [comment]
I just checked better and LMD files do not actually have a TIMESTEP
parameter, not a corresponding one. However, the first segment of LMD files
contain the time in seconds, hence it is still possible to re-build the
flow rate.
I have now uploaded a package in the development branch of bioconductor
which should be able to analyse properly also the LMD files. The version is
1.71.1
You could download it here for now:
https://www.dropbox.com/s/cmknqsjo3ofjg1s/flowAI_1.17.1.tar.gz?dl=0

Let me know if it works.

Best wishes,
Gianni

**Comment by jgarces02** [comment]
Hi @giannimonaco, sorry for not answering your so dedicated answer... I've put a bit apart (some time, I hope not so much) this project but a friend of me, running this function **with normal FCS files**, got this error: 
```
> physical.markers <- c("FSC_A", "FSC_H", "SSC_A", "SSC_H")
> idx <- flow_auto_qc(fcs, ChExcludeFS =  physical.markers, fcs_QC = FALSE, output = 3, html_report = F, folder_results = F, mini_report = F)
Error in flowAI::flow_auto_qc(fcs[[file]], ChExcludeFS = physical.markers, ...)
object "FileType" not found
```
Digging into the code, we've found that maybe the error could be in the section you added to read LMD files:
```
if( is.character(fcsfiles) ){
    FileType <- toupper(strsplit(basename(fcsfiles[1]), split="\\.")[[1]][-1])
    if(FileType == "LMD"){
      set <- read.flowSet(files = fcsfiles, dataset = 2)
    }else{
      set <- read.flowSet(files = fcsfiles) 
    }
```
Thanks again for your help and I'd like go deeper with LMD files in some time... I'd notify you how this new adding works!

**Comment by giannimonaco** [comment]
Hi, thank you for letting me know about the issue. It seems that the files
of your friend do not have an extension. Is that right?
I added a checking step to verify if files have an extension and if not
flowAI proceed analyzing fcs files.

Also note that I added the repository in GitHub:
https://github.com/giannimonaco/flowAI
This repository is syncronized with the one in bioconductor!
